### PR TITLE
Position node autocomplete UI against respective input port

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

https://jira.autodesk.com/browse/DYN-3174

Attempt at placing the node autocomplete UI aligned with the respective port. 

This works as expected **except for the first time you bring up the UI**, for some reason. As can be seen in the gif, the first time, the UI is not spaced correctly from the port. It was observed that the reason for this is that the `ActualWidth` of the control is 250 the first time and subsequently it is 256. I'm not sure why this is the case 😕 

place_autocomplete_ui

### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### FYIs

@Amoursol

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/Search/NodeSearchElementViewModel.cs [hunk 1]
```diff
@@ -290,10 +290,7 @@ protected virtual void CreateAndConnectToPort(object parameter)
             DynamoSelection.Instance.ClearSelection();
             var inputNodes = initialNode.InputNodes.Values.Where(x => x != null).Select(y => y.Item2);
 
-            foreach (var inputNode in inputNodes)
-            {
-                DynamoSelection.Instance.Selection.AddUnique(inputNode);
-            }
+            DynamoSelection.Instance.Selection.AddRange(inputNodes);
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Search/NodeSearchElementViewModel.cs:293]: 👍

File: src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml [hunk 2]
```diff
@@ -349,8 +349,6 @@
                StaysOpen="True"
                AllowsTransparency="True"
                IsOpen="False"
-               Placement="MousePoint"
-               HorizontalOffset="-290"
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml:353]: Cool, was expecting these removed

File: src/DynamoCoreWpf/ViewModels/Search/NodeAutoCompleteSearchViewModel.cs [hunk 3]
```diff
@@ -13,6 +12,7 @@ public class NodeAutoCompleteSearchViewModel : SearchViewModel
     {
         internal PortViewModel PortViewModel { get; set; }
 
+
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Search/NodeAutoCompleteSearchViewModel.cs:15]: extra line?

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs [hunk 4]
```diff
@@ -226,6 +228,27 @@ public override void Dispose()
             _node.WorkspaceViewModel.PropertyChanged -= Workspace_PropertyChanged;
         }
 
+        internal void PlaceNodeAutocompleteWindow(object sender, EventArgs e)
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:231]: @aparajit-pratap Would you add some comments for this function? and maybe in the comments also mention how to adjust the offset in the future

File: src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs [hunk 5]
```diff
@@ -155,6 +155,7 @@ void OnWorkspaceViewUnloaded(object sender, RoutedEventArgs e)
             if (ViewModel != null)
             {
                 removeViewModelsubscriptions(ViewModel);
+                ViewModel.RequestNodeAutoCompleteSearch -= ShowHideNodeAutoCompleteControl;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs:158]: Not sure how I missed this, but this line can maybe placed as part of `removeViewModelsubscriptions()`

File: src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs [hunk 6]
```diff
@@ -186,6 +187,7 @@ private void ShowHidePopup(ShowHideFlags flag, Popup popup)
                 case ShowHideFlags.Show:
                     // Show InCanvas search just in case, when mouse is over workspace.
                     popup.IsOpen = DynamoModel.IsTestMode || IsMouseOver;
+
```
Comment (QilongTang) [inline src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs:190]: extra line here

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs [hunk 7]
```diff
@@ -226,6 +228,27 @@ public override void Dispose()
...
+                // Position node autocomplete UI offset left from X position of node by its width.
+                x = _node.X - popup.ActualWidth;
+            }
+            else
+            {
+                // Position node autocomplete UI offset right from X position of node by node width.
+                x = _node.X + _node.NodeModel.Width;
+            }
+            // Position UI down from the top of the node but offset by the node header and against the respective port.
+            var y = _node.Y + NodeModel.NodeHeaderHeight + PortModel.Index * PortModel.Height;
```
Comment (QilongTang) [inline src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:247]: This way I guess there is no space between?

File: src/DynamoCore/Graph/Nodes/PortModel.cs [hunk 8]
```diff
@@ -163,10 +163,9 @@ public Point2D Center
             get
             {
                 double halfHeight = Height * 0.5;
-                const double headerHeight = 25;
 
                 double offset = Owner.GetPortVerticalOffset(this);
-                double y = Owner.Y + headerHeight + 5 + halfHeight + offset;
+                double y = Owner.Y + NodeModel.NodeHeaderHeight + 5 + halfHeight + offset;
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Nodes/PortModel.cs:168]: 👍

File: src/DynamoCore/Graph/Nodes/NodeModel.cs [hunk 9]
```diff
@@ -64,12 +64,14 @@ public abstract class NodeModel : ModelBase, IRenderPackageSource<NodeModel>, ID
         private ObservableCollection<PortModel> inPorts = new ObservableCollection<PortModel>();
         private ObservableCollection<PortModel> outPorts = new ObservableCollection<PortModel>();
 
+        private readonly Dictionary<int, Tuple<int, NodeModel>> inputNodes;
+        private readonly Dictionary<int, HashSet<Tuple<int, NodeModel>>> outputNodes;
+
         #endregion
 
-        #region public members
+        internal const double NodeHeaderHeight = 25;
```
Comment (QilongTang) [inline src/DynamoCore/Graph/Nodes/NodeModel.cs:72]: Given the X, Y are already obsoleted, it seems strange we keep adding this on NodeModel level, is there other options?

Comment (aparajit-pratap) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:72]: This is also needed in `portmodel` so if I add it to `NodeViewModel`, it won't be available to the `portmodel`.

File: src/DynamoCore/Graph/Nodes/PortModel.cs (hunk 8 above)
Comment (aparajit-pratap) [reply src/DynamoCore/Graph/Nodes/PortModel.cs:168]: Will rename.

File: src/DynamoCore/Graph/Nodes/NodeModel.cs (hunk 9 above)
Comment (aparajit-pratap) [reply src/DynamoCore/Graph/Nodes/NodeModel.cs:72]: Also, I don't think we can obsolete X, Y. From my understanding, they are still necessary and even the `NodeViewModel`, X, and Y properties eventually derive from those of the `NodeModel`.

File: src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs (hunk 7 above)
Comment (aparajit-pratap) [reply src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:247]: No, there is no space. I can add some constant spacing, just wasn't sure the exact amount. Will add a GIF for folks to review.

**Comment by Amoursol** [comment]
@aparajit-pratap looking great! 🙌 

Can we simply change the initializing width to be 256 by default, then pick up actual width? Or is this only in one isolated case and could be different for other ones?

**Review by QilongTang — APPROVED** [review]
LGTM

**Comment by QilongTang** [comment]
@aparajit-pratap Would you try https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml#L10? They may end up both working but I think limiting on the control itself may be cleaner

**Review by QilongTang — APPROVED** [review]
@aparajit-pratap LGTM

**Pull request merged**